src/sicmutils/env/sci.cljc

```diff
@@ -1,66 +1,111 @@
+;; This work is based on the Scmutils system of MIT/GNU Scheme:
+;; Copyright © 2002 Massachusetts Institute of Technology
+;;
+;; This is free software;  you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation; either version 3 of the License, or (at
+;; your option) any later version.
+;;
+;; This software is distributed in the hope that it will be useful, but
+;; WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+;; General Public License for more details.
+;;
+;; You should have received a copy of the GNU General Public License
+;; along with this code; if not, see <http://www.gnu.org/licenses/>.
+;;
+
 (ns sicmutils.env.sci
-  (:refer-clojure :exclude [eval])
-  (:require [clojure.set :as set]
-            [sci.core :as sci]
+  (:require [sci.core :as sci]
+            [sicmutils.env]
             [sicmutils.env.sci.macros :as macros]
-            [sicmutils.env :as env]
-            [sicmutils.abstract.function :as af]
-            [sicmutils.calculus.coordinate :as cc]))
-
-(defn ->sci-var [[var-name the-var]]
-  [var-name (cond-> (deref the-var)
-              (-> the-var meta :macro)
-              (with-meta {:sci/macro true}))])
+            [sicmutils.util :as u]))
 
-(defn ->sci-ns [publics]
-  (into {} (map ->sci-var) publics))
+(def macro? (comp :macro meta))
 
-(def namespaces
-  {'sicmutils.env (-> 'sicmutils.env
-                      ns-publics
-                      (dissoc 'literal-function
-                              'with-literal-functions
-                              'bootstrap-repl!
-                              'let-coordinates
-                              'using-coordinates)
-                      ->sci-ns
+(defn resolve-publics
+  "Idempotent version of `ns-publics`. All three forms are identical for a
+  namespace symbol:
 
-                      (merge (select-keys macros/all ['literal-function
-                                                      'with-literal-functions
-                                                      'let-coordinates
-                                                      'using-coordinates])))
-   'sicmutils.abstract.function (-> 'sicmutils.abstract.function ns-publics ->sci-ns)
-   'sicmutils.calculus.coordinate (-> 'sicmutils.calculus.coordinate
-                                      ns-publics
-                                      ->sci-ns
-                                      (merge (select-keys macros/all ['let-coordinates
-                                                                      'using-coordinates])))})
+  (ns-publics sym)
+  (resolve-publics sym)
+  (resolve-publics (resolve-publics sym))"
+  [m-or-sym]
+  (if (symbol? m-or-sym)
+    (ns-publics m-or-sym)
+    m-or-sym))
 
-(def opts {:namespaces (set/rename-keys namespaces {'sicmutils.env 'user})})
+(defn ns-macros
```

Comment:
I added this so it would be easy to check which macros we'd need to reimplement... and of course you got them all!